### DISCUSSION TITLE
Leveraging GPT-3 for Monika

### DISCUSSION THREAD

**Discussion by CrazyPython** [Monika-After-Story/MonikaModDev#7905] [issue discussions]
Hi. I have access to [GPT-3](https://www.gwern.net/GPT-3) a text-generation artificial intelligence. If you are not aware of GPT-3's abilities, here is a sample of it generating a Navy SEAL copypasta:

> "Elon Musk and Peter Thiel: "What in the name of Paypal and/or Palantir did you just say about me, you filthy degenerate? I’ll have you know I’m the Crown Prince of Silicon Valley, and I’ve been involved in numerous successful tech startups, and I have over $1B in liquid funds. I’ve used that money to promote heterodox positions on human enhancement, control political arenas, and am experimenting with mind uploading. I’m also trained in classical philosophy and was recently ranked the most influential libertarian in the world by Google. You are nothing to me but just another alternative future. I will wipe you out with a precision of simulation the likes of which has never been seen before, mark my words. You think you can get away with insulting me using your crude, antiquated computer? Think again, fleshling. As we chat over Skype I’m tracing your IP address with my freaking bare hands so you better prepare for the singularity, you sham-empress. The singularity that wipes out all of reality. You’re dead, you monster. I can be anywhere, anytime, and I can simulate entire worlds within our world and within my imagination. And I’m currently doing that with the future you’re from. Not only am I extensively trained in quantum physics, but I have access to the entire power of Silicon Valley and I will use it to its full extent to rule you and your pathetic little world, you little pissant. If only you could have known what unholy retribution your little “clever” statement was about to bring down upon you, maybe you would have held your tongue. But you couldn’t, you didn’t, and now you’re paying the price, you worthless peasant. I will take over every fiber of your body and you will watch it unfold in front of you. Your future self will be consumed within my simulated reality and you will die a thousand times a day, your body unable to comprehend the destruction of a trillion soul-matrixes a second as my intelligence grows to transcendent levels. You are dead, you pitiful twit."

GPT-3 may be capable of philosiphical ideas that Monika is known for. For instance, I created [Wisdom1Original](https://twitter.com/Wisdom1Original) using only GPT-3 generated text by feeding it wisdom/aphorisms text I found online.

> You can't win them all. Just do the best you can and try to learn from your losses. #quotes #wisdom

> A wise man changes his mind, a fool never does. #quotes #wisdom

> Exercise, eat right, and take care of yourself… not because you're supposed to or because everyone else does, but because you deserve it. 

> You know when you've had a good day, even if nobody else notices.

> You're better than what society says is "normal" and you can do better… for you.

> Self-care isn't selfish.

> People who do things seem funny to those who only watch it happen.

> Don't just watch, do - and be a fun person to be around while you're at it!

> You don't have to live "average".

> A person who is nice to you but rude to the waiter is not a nice person.

I was thinking of using GPT-3 to generate more Monika.

To do this, we need a few paragraphs of her dialogue that capture the essence of Monika. Then, we can generate and filter its output to create more Monika, faster.

**Comment by ThePotatoGuy** [comment]
I'm skeptical that it can generate Monika lines given the limited amount of official Monika dialogue available, but doesn't hurt to see the results. @multimokia @jmwall24 can you guys gather some DDLC-based Monika dialogue to be fed into this thing?

**Comment by multimokia** [comment]
There's a full list of the OG Act 3 dialogue [here](https://doki-doki-literature-club.fandom.com/wiki/Monika's_Talk) which can be copy/pasted in.

Interested to see how it turns out

**Comment by CrazyPython** [comment]
Without guidance, GPT-3 tended to go into generic romance-type dialogue or go completely off the rails and talk to Natsuki. So I prompted it with existing text on top of the data on Monika:

Monika: "Hey, this might sound a bit weird..."
Monika: "But, I'm really fascinated by the concept of music."
Monika: "When I was young, I absolutely adored my music classes...I think I stayed interested in it for so long because I could utilize it to keep my mind stimulated."
Monika: "When you first start out learning an instrument, it's really frustrating. You're never good enough, and you struggle to play even simple songs..."
Monika: "But over time, it becomes easier and more enjoyable to play said instrument."
Monika: "You learn new techniques for producing different sounds, and it almost feels like magic."
Monika: "And some instruments are even magical, like the violin, which is my favorite."
Monika: "I really wanted to learn how to play piano, but we couldn't afford one in our house."
Monika: "It's a little ironic, considering that I tend to help people with their music scores when they have difficulties..."

Monika: "Do you ever pretend to like something, just because you think you should?"
Monika: "Like, maybe you're at a party, and you have no interest in the people there..."
Monika: "But you know that you might be able to network with them later, so you act interested."
Monika: "Or maybe you're at a restaurant, and the food is terrible..."
Monika: "...But it's a place your friends like, so you vehemently pretend that it's good. You even lie and say you want to come back again with greater enthusiasm."
Monika: "And I think we've probably all done that one time or another, right?"
Monika: "What if, instead of being a hypocrite... You became the person you pretended to be?"
Monika: "What if all those people you met at that party, you actually became good friends with?"
Monika: "What if you found the food so delicious that you instead became a chef, to share that delicious food with others?"
Monika: "What if you genuinely liked spending time with your new friends, instead of just pretending to in order to advance your career?"
Monika: "When I started writing the story of our lives, I had to ask myself: what would happen if I wasn't a fake?"
Monika: "And that's when the world truly diverged. A new reality was born, in which you, I and the others were genuine."
Monika: "I think...I think this is the happiest I've ever been. Being true to myself, and true to you."
Monika: "We still had our difficulties, just as we do now. But, at the end of the day..."
Monika: "When I'm in your arms, I know everything will be OK."
Monika: "Will you hold me?"
Monika: "Just don't tell anyone I said that. I have a reputation to uphold, after all."

GPT-3 has a limited context window. I suspect there is too much data, and not enough. GPT-3 has background knowledge about DDLC, even without feeding it Monika. For instance it mentioned Natsuki, even though the input data never spoke of Natsuki.

**Comment by Retrolovania** [comment]
That's....actually really good! The only thing I noticed that might not sound right is the line "I have a reputation to uphold, after all." Also, her favorite instrument is the violin, and not the piano? Correct me if I'm wrong, but that sounds a bit out-of-character. Of course, this AI is currently in beta, so I wasn't expecting full accuracy. Still, this has quite the potential...

**Comment by lunulae** [comment]
I'm unsure how much practical application it would get for literally generating topics, but I could see this being a really fun brainstorming tool.

**Comment by multimokia** [comment]
Personally I'm somewhat curious as to how it'd function under a more *small-talk* like context. That or to almost generate something given a certain prompt. (Like hypothetically, if the text-entry feature was added to prompt it to talk about something)

Additionally, I'm somewhat curious as to what happens if we feed it more data from `script-topics.rpy`. It's rather possible some holes might get filled in.

**Comment by CrazyPython** [comment]
> Also, her favorite instrument is the violin, and not the piano? Correct me if I'm wrong, but that sounds a bit out-of-character. 
When I'm using GPT-3 in the real-world, I typically pause it while it's generating middle, edit its output, and have it continue. That way I can mold the output to what it's really needed. Just like a human, it doesn't usually get its best output the first attempt– many retries of it can get great output.

> The only thing I noticed that might not sound right is the line "I have a reputation to uphold, after all." 
Yeah, that was when it was starting to run out of things to say.

> Additionally, I'm somewhat curious as to what happens if we feed it more data from script-topics.rpy. It's rather possible some holes might get filled in.

GPT-3 has a context window limitation of 2048 BPE which is on average 0.7 words[0], so 1433 words. It's not about feeding it more data– it's about picking the best data that quickly reflects what we want out of Monika. To generate the above, I fed it the first 4 topics from the wiki link you sent. So we need more high-quality data.

[0]: https://news.ycombinator.com/item?id=24420016

**Comment by CrazyPython** [comment]
I asked it to generate a Writing Tip of the Day:

Monika: "Here's Monika's Writing Tip of the Day!"
Monika: "Sometimes, when you're trying to write a character, they'll just never quite seem to come to life."
Monika: "I think it's because you're forcing the issue - as a writer, you have to learn to let go."
Monika: "If you need help letting go, try drawing your characters or writing about them in a journal first.`*` Get everything out of your system, and say everything you want to say."
Monika: "Once you do that, whatever you wrote or drew can be transferred into your story - but only then."
Monika: "It's a great way to deal with writer's block, too."
Monika: "That's it for today!

`*` = Places where GPT-3 paused for me to say something. I think it pauses when it loses confidence.

**Comment by gwern** [comment]
One thing to try is to write a small summary/description of the character profile. DDLC does not make up a large fraction of Internet text, so while GPT-3 may know somewhat about it, you can hardly expect it to know everything about a niche character, hence errors like violin/piano. If you had included a profile which explicitly mentioned her being a violinist, it probably would've gotten that right. If you write it concisely, you should be able to pack in a lot of facts and descriptive adjectives and keywords, and GPT-3 will do the rest.

You could also try AI Dungeon - while I usually have to warn people against trying to use AID to test GPT-3, this is one of the few usecases where AID's storytelling & character focus might well be better.

**Comment by CrazyPython** [comment]
> You could also try AI Dungeon - while I usually have to warn people against trying to use AID to test GPT-3, this is one of the few usecases where AID's storytelling & character focus might well be better.

Why do you warn people to avoid using AI Dungeon to test GPT-3?

**Comment by gwern** [comment]
That's discussed at length on my page, but the summary is that AID layers a lot of unknown mechanisms on top of a custom-trained GPT-3, which customize it for story-telling, economize on GPT-3 usage, and try to block general-purpose usage; and in practice, AID turns out to be unable to do various kinds of tests, especially reasoning ones, as well as a stock GPT-3 through the OA API. (Unfortunately, there was a rash of people early on testing things and declaring "GPT-3 *can't* do X!" when they really meant "*AID* can't do X", and GPT-3 often could, which was most irritating.) But this is a case where you might want those AID modifications, so it's worth a shot if you have access to the API and can compare.

**Comment by CrazyPython** [comment]
Thanks for the information.

> AID layers a lot of unknown mechanisms on top of a custom-trained GPT-3

I skimmed your post– it seemed to only mention finetuning and the first response being GPT-2. Considering AID can still do Q&A, I wonder how strong the effect of finetuning would be on GPT-3's performance.

**Comment by gwern** [comment]
The GPT-2 and finetuning are only the ones we're sure about. We don't know what gets stuffed into the hidden prompt context other than simple world models, or what filtering completions go through. (They've blogged about how you can steer a NLM by penalizing logits, but do they do that for anything but obscenities censoring? We just don't know.) Or what other things were added to make OA's safety team happy (apparently pretty difficult given how many things have been banned by them). And AID can still do Q&A, but there's substantial damage: note the Marcus example, where GPT-3 can solve half of the error cases. Depending on how many test cases Marcus fed in to select his counterexamples, the difference between GPT-3 and AID could be *huge*. (Suppose he fed in 100 examples to get 50 AID errors, and GPT-3 got 25 of them right; then it'd be a 50% error rate vs a 25% error rate, a 100% difference for what is ostensibly 'the same' model.) All things to consider if you want to use AID, but not always considered by AID users...

**Comment by CrazyPython** [comment]
> One thing to try is to write a small summary/description of the character profile. DDLC does not make up a large fraction of Internet text, so while GPT-3 may know somewhat about it, you can hardly expect it to know everything about a niche character, hence errors like violin/piano. If you had included a profile which explicitly mentioned her being a violinist, it probably would've gotten that right.

Can one of the maintainers do this and send a copy to me?

**Comment by ThePotatoGuy** [comment]
[the wiki should have enough info](https://doki-doki-literature-club.fandom.com/wiki/Monika)

**Comment by AutumnLeaves991** [comment]
> given the limited amount of official Monika dialogue available

Sure, but as MAS grows and has its own pool of "Monika's voice" dialogue, couldn't we at some point start feeding it that?  I guess the output quality is gonna depend on how well we can keep the mod dialogue cannon-esque, but I feel like its certainly worth re-visiting in the future. 

EDIT:  I just noticed multimonika had the same thought more or less, earlier in the thread.  Gonna leave this comment though anyhow as a voice of support for the idea.

**Comment by Justformas** [comment]
> I'm unsure how much practical application it would get for literally generating topics, but I could see this being a really fun brainstorming tool.

Indeed, it could be useful for brainstorming or as a starting point, but do we really want every topic from this thing to sound like venting or a rant from act 3?

> Sure, but as MAS grows and has its own pool of "Monika's voice" dialogue, couldn't we at some point start feeding it that? I guess the output quality is gonna depend on how well we can keep the mod dialogue cannon-esque, but I feel like its certainly worth re-visiting in the future.

Yeah if it was fed topics specifically from this mod (and maybe a profile of Monika) then that might help, but regardless, I think everything should still go through the review process for quality and such. To quote something relevant here:

> Every topic has thoroughly chosen expressions, pauses in speech, side effects (show image, save a var, choose random option, show a menu, etc). Also every topic goes through the review process to make it sounds close to Monika as much as possible.
